### PR TITLE
Update 2.1.0: CPU only

### PR DESCRIPTION
**Pull request by danpetry**
CPU only, GPU to follow later.

Upstream: https://github.com/pytorch/pytorch

Changes:
- build arm-64 on ventura machine to support GPU variant
- manually copy symlinked files to solve windows build issue
- tidy and correct the gpu variant env variable handling in setup.py:
  - turn off mps when not used
  - configure blas alongside GPU stuff, not instead of, as you can have blas and cuda/mps concurrently
- add pytorch-cpu and pytorch-gpu metapackages, as well as automatic installation of cpu or gpu variants according to platform support, if the pytorch package is installed
- update patches

### CODE REVIEW

File: recipe/bld_pytorch.bat
Comment (danpetry) [inline recipe/bld_pytorch.bat:54]:
```suggestion
```

Comment (danpetry) [reply recipe/bld_pytorch.bat:54]: something I noticed, will merge after review

File: recipe/bld_pytorch.bat
Comment (danpetry) [inline recipe/bld_pytorch.bat:53]:
```suggestion
```

File: recipe/bld_pytorch.bat
Comment (danpetry) [inline recipe/bld_pytorch.bat:56]:
```suggestion
```

File: recipe/build_pytorch.sh
Comment (danpetry) [inline recipe/build_pytorch.sh:133]:
```suggestion
            export TORCH_CUDA_ARCH_LIST="3.5+PTX;5.0;6.0;7.0"
```

File: recipe/build_pytorch.sh
Comment (danpetry) [inline recipe/build_pytorch.sh:135]:
```suggestion
            export TORCH_CUDA_ARCH_LIST="3.5+PTX;5.0;6.0;6.1;7.0"
```

File: recipe/build_pytorch.sh
Comment (danpetry) [inline recipe/build_pytorch.sh:137]:
```suggestion
            export TORCH_CUDA_ARCH_LIST="3.5+PTX;5.0;6.0;6.1;7.0;7.5"
```

File: recipe/build_pytorch.sh
Comment (danpetry) [inline recipe/build_pytorch.sh:139]:
```suggestion
            export TORCH_CUDA_ARCH_LIST="3.5+PTX;5.0;6.0;6.1;7.0;7.5;8.0"
```

File: recipe/build_pytorch.sh
Comment (danpetry) [inline recipe/build_pytorch.sh:141]:
```suggestion
            export TORCH_CUDA_ARCH_LIST="3.5+PTX;5.0;6.0;6.1;7.0;7.5;8.0;8.6"
```

File: recipe/build_pytorch.sh
Comment (danpetry) [inline recipe/build_pytorch.sh:143]:
```suggestion
            export TORCH_CUDA_ARCH_LIST="3.5+PTX;5.0;6.0;6.1;7.0;7.5;8.0;8.6"
```

File: recipe/build_pytorch.sh
Comment (danpetry) [inline recipe/build_pytorch.sh:145]:
```suggestion
            export TORCH_CUDA_ARCH_LIST="3.5+PTX;5.0;6.0;6.1;7.0;7.5;8.0;8.6"
```

File: recipe/bld_pytorch.bat
Comment (cbouss) [inline recipe/bld_pytorch.bat:8]: Is this needed when building from the release archive? Or is it just when building from a git checkout?
If the later, you could put this under a condition by setting a env var when rc is not None in meta.yaml

File: abs.yaml
Comment (cbouss) [inline abs.yaml:7]:
```suggestion
  - "--skip-existing"
```
Minor comment, if you do end up committing changes, skip-existing would be preferred.

File: recipe/bld_pytorch.bat
Comment (danpetry) [reply recipe/bld_pytorch.bat:8]: no, it's needed when building from the release archive

File: abs.yaml
Comment (danpetry) [reply abs.yaml:7]: do you mean setting skip CI? Nothing's uploaded, or will be uploaded before the PR's merged, so AFAIU skip-existing wouldn't do much?

File: abs.yaml
Comment (M-Waszkiewicz-Anaconda) [inline abs.yaml:12]: does it mean we only build on ventura ?

File: recipe/meta.yaml
Comment (M-Waszkiewicz-Anaconda) [inline recipe/meta.yaml:56]: what is special about py 310?

**Review by rafaelmartins — APPROVED** [review]
looks ok

File: abs.yaml
Comment (cbouss) [reply abs.yaml:7]: not having skip-existing is good when you configure a dev channel, but gives you the risk of uploading an existing version when it is not set.

Comment (danpetry) [reply abs.yaml:7]: added back

File: abs.yaml
Comment (JeanChristopheMorinPerso) [reply abs.yaml:12]: Yes this is what it means.

File: recipe/bld_pytorch.bat
Comment (danpetry) [reply recipe/bld_pytorch.bat:59]: changed, thanks

Comment (JeanChristopheMorinPerso) [reply recipe/bld_pytorch.bat:59]: I think we limit to 2 because of the amount of RAM used.

File: recipe/build_pytorch.sh
Comment (JeanChristopheMorinPerso) [reply recipe/build_pytorch.sh:24]: This is the standard path, so no.

File: recipe/meta.yaml
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml:13]: Could we use `https://github.com/pytorch/pytorch/archive/refs/tags/v{{ version }}-rc{{ rc }}.tar.gz` to download the release candidates instead of using git?

File: recipe/build_pytorch.sh
Comment (danpetry) [reply recipe/build_pytorch.sh:24]: [Here's](https://anaconda.slack.com/archives/C03J3NQPS0Z/p1685646003488849?thread_ts=1683227605.879399&cid=C03J3NQPS0Z) the conversation about this. It helps with finding the SDK on the CI. Local builds still work fine. (This directory exists on my osx-arm64 machine at least.) AFAIU, the [script_env](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#use-environment-variables) section is to pass through environment variables from the command line, which isn't something we need/want to do here. Or could you clarify, if I'm misunderstanding?

File: recipe/meta.yaml
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:56]: Or more, what is special with 39 and 311. Because right now it'll only compile on 310.

File: abs.yaml
Comment (danpetry) [reply abs.yaml:12]: yes, but since the SDK used for CPU is older, it will work on earlier macs

File: recipe/meta.yaml
Comment (danpetry) [reply recipe/meta.yaml:56]: thanks, good catch

Comment (danpetry) [reply recipe/meta.yaml:56]: removed

Comment (danpetry) [reply recipe/meta.yaml:56]: this was there to reduce development time during testing

**Comment by danpetry** [comment]
The builds here were done with python 3.10 only (by accident). They will most likely work with other pythons too, but I'd rather save another 1-day build cycle. If reviewers are ok with approving based on these build results, that would be helpful

File: recipe/meta.yaml
Comment (danpetry) [reply recipe/meta.yaml:13]: yeah, I tried that originally and then changed it to this method. My commit messages don't say why I did it, but I suspect that it's because the refs/tags tarball doesn't contain the submodules

**Comment by danpetry** [comment]
@M-Waszkiewicz-Anaconda @JeanChristopheMorinPerso @cbouss if I've addressed all your comments, is an approval possible?

File: recipe/bld_pytorch.bat
Comment (danpetry) [reply recipe/bld_pytorch.bat:60]: Oh yes. It runs out of heap space if we increase it too hard. It was hard-set [here](https://github.com/AnacondaRecipes/pytorch-feedstock/commit/f643db0c7302bdd9690f69f54c21fc48b1f90019) and subsequently decreased

**Review by JeanChristopheMorinPerso — APPROVED** [review]
Great work!

**Comment by danpetry** [comment]
Thanks JC :)

File: recipe/meta.yaml
Comment (danpetry) [reply recipe/meta.yaml:13]: the official releases do, but the archive/refs/tags tarballs is just a github-provided feature and only contains the repo contents

Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:13]: Yep. What I was saying is that it's annoying that GH's default archives don't include submodules.

**Pull request merged**
